ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/PullPhysicalPlanBuilder.java

```diff
@@ -0,0 +1,348 @@
+    LogicalSchema outputSchema = null;
+    boolean isStar = false;
+    if (isSelectStar(statement.getStatement().getSelect())) {
+      isStar = true;
+      outputSchema = buildSchema(mat.schema(), mat.windowType().isPresent());
+    } else {
+      final List<SelectExpression> projection = analysis.getSelectItems().stream()
+          .map(SingleColumn.class::cast)
+          .map(si -> SelectExpression
+              .of(si.getAlias().orElseThrow(IllegalStateException::new), si.getExpression()))
+          .collect(Collectors.toList());
+
+      outputSchema = selectOutputSchema(
+          executionContext, projection, mat.windowType());
+    }
+    return new ProjectOperator(
+      config,
+      metaStore,
+      mat,
+      analysis,
+      executionContext,
+      contextStacker,
+      logicalNode,
+      outputSchema,
+      isStar);
+  }
+
+  private SelectOperator translateFilterNode(final FilterNode logicalNode) {
+    whereInfo = WhereInfo.extractWhereInfo(analysis, persistentQueryMetadata);
+    return new SelectOperator(logicalNode);
+  }
+
+  private AbstractPhysicalOperator translateDataSourceNode(
+      final DataSourceNode logicalNode,
+      final PersistentQueryMetadata persistentQueryMetadata
+  ) {
+    if (whereInfo == null) {
+      throw new KsqlException("Pull queries must have a WHERE clause");
+    }
+    keys = whereInfo.getKeysBound().stream()
+        .map(keyBound -> asKeyStruct(keyBound, persistentQueryMetadata.getPhysicalSchema()))
+        .collect(ImmutableList.toImmutableList());
+
+    if (!whereInfo.getWindowBounds().isPresent()) {
+      return new KeyedTableLookupOperator(mat, logicalNode);
+    } else {
+      return new KeyedWindowedTableLookupOperator(
+          mat, logicalNode, whereInfo.getWindowBounds().get());
+    }
+  }
+
+  private PersistentQueryMetadata findMaterializingQuery(
+      final KsqlExecutionContext executionContext,
+      final ImmutableAnalysis analysis
+  ) {
+    final MetaStore metaStore = executionContext.getMetaStore();
+
+    final SourceName sourceName = getSourceName(analysis);
+
+    final Set<String> queries = metaStore.getQueriesWithSink(sourceName);
+    if (queries.isEmpty()) {
+      throw notMaterializedException(sourceName);
+    }
+    if (queries.size() > 1) {
+      throw new KsqlException(
+        "Multiple queries currently materialize '" + sourceName + "'."
+        + " KSQL currently only supports pull queries when the table has only been"
+        + " materialized once.");
+    }
+
+    final QueryId queryId = new QueryId(Iterables.get(queries, 0));
+
+    final PersistentQueryMetadata query = executionContext
+        .getPersistentQuery(queryId)
+        .orElseThrow(() -> new KsqlException("Materializing query has been stopped"));
+
+    if (query.getDataSourceType() != DataSourceType.KTABLE) {
+      throw new KsqlException("Pull queries are not supported on streams.");
```

Comment:
Could we also include the actual `getDataSourceType` as well.

Also, for other thrown exceptions, I find that we tend to not include some key information into the exception message that may make debugging harder. I'd suggest as a rule of thumb to always include all variable values if they participate in the condition that leads to exception.